[0.43.11-RC5 : Bug] Fatal error when "Restore Unit" Black Eagle

### Environment

0.43.11-RC5

### Description

Fatal error when "Restore Unit" Black Eagle. Game will be frozen. Nothing in log. Might be related to case #745 and #747. To replicate:

1. Load campaign
2. Go to hanger
3. Right click, GM Mode, Restore Unit
4. Game frozen

![image](https://user-images.githubusercontent.com/17064665/41187058-eb6b8288-6bd4-11e8-9d30-1b81b99d3231.png)

### Files

[Fatal error when Restore Unit Black Eagle.zip](https://github.com/MegaMek/mekhq/files/2086588/Fatal.error.when.Restore.Unit.Black.Eagle.zip)

The InvocationTargetException in NativeMethodAccessorImpl.invoke0 is a wrapper around a different exception thrown in the process of dispatching an event. The cause is the exception thrown by the event listener, which in this case is a NullPointerException when attempting to update the daily log. The stack trace from the original exception is truncated, but this is what is in the log:

```Caused by: java.lang.NullPointerException
	at mekhq.gui.DailyReportLogPanel.appendLog(DailyReportLogPanel.java:112)
	at mekhq.gui.CampaignGUI.refreshReport(CampaignGUI.java:2969)
	at mekhq.gui.CampaignGUI.handle(CampaignGUI.java:3029)
	... 48 more```

I was unable to replicate it and there is nothing at that point in the code that suggests a cause to me.

--- a/MekHQ/src/mekhq/campaign/unit/Unit.java
+++ b/MekHQ/src/mekhq/campaign/unit/Unit.java
@@ -4921,8 +4921,24 @@ public class Unit implements MekHqXmlSerializable, ITechnology {
         return en.getDamageLevel(false);
     }
 
-    public void resetParts() {
-        parts = new ArrayList<>();
+    /**
+     * Removes all of the parts from a unit.
+     * 
+     * NOTE: this puts the unit in an inconsistent state, and
+     *       the unit should not be used until its parts have
+     *       been re-assigned.
+     * 
+     */
+    public void removeParts() {
+        for (Part part : parts) {
+            part.setUnit(null);
+
+            if (campaign != null) {
+                campaign.getWarehouse().removePart(part);
+            }
+        }
+
+        parts.clear();
     }
 
     /**

--- a/MekHQ/src/mekhq/campaign/unit/actions/RestoreUnitAction.java
+++ b/MekHQ/src/mekhq/campaign/unit/actions/RestoreUnitAction.java
@@ -184,7 +184,7 @@ public class RestoreUnitAction implements IUnitAction {
     public interface IEntityCopyFactory {
         /**
          * Gets a copy of the entity.
-         * @param en The entity to copy.
+         * @param entity The entity to copy.
          * @return A copy of the entity, or {@code null} if a copy could not be made.
          */
         @Nullable
@@ -198,18 +198,18 @@ public class RestoreUnitAction implements IUnitAction {
     private static class FileSystemEntityCopyFactory implements IEntityCopyFactory {
         /**
          * Get a copy of the entity from the {@link MechSummaryCache}.
-         * @param en The entity to copy.
+         * @param entity The entity to copy.
          * @return A copy of the entity, or {@code null} if a copy could not be made.
          */
         @Nullable
-        public Entity copy(Entity en) {
-            final MechSummary ms = MechSummaryCache.getInstance().getMech(en.getShortNameRaw());
+        public Entity copy(Entity entity) {
+            final MechSummary ms = MechSummaryCache.getInstance().getMech(entity.getShortNameRaw());
             try {
                 if (ms != null) {
                     return new MechFileParser(ms.getSourceFile(), ms.getEntryName()).getEntity();
                 }
             } catch (EntityLoadingException e) {
-                MekHQ.getLogger().error("Cannot restore unit from entity, could not find: " + ms.getName(), e);
+                MekHQ.getLogger().error("Cannot restore unit from entity, could not find: " + entity.getShortNameRaw(), e);
             }
 
             return null;

--- a/MekHQ/src/mekhq/campaign/unit/actions/RestoreUnitAction.java
+++ b/MekHQ/src/mekhq/campaign/unit/actions/RestoreUnitAction.java
@@ -22,6 +22,8 @@ package mekhq.campaign.unit.actions;
 import java.util.*;
 
 import megamek.common.*;
+import megamek.common.annotations.Nullable;
+import megamek.common.loaders.EntityLoadingException;
 import mekhq.MekHQ;
 import mekhq.campaign.Campaign;
 import mekhq.campaign.event.UnitChangedEvent;
@@ -34,8 +36,87 @@ import mekhq.campaign.unit.Unit;
  */
 public class RestoreUnitAction implements IUnitAction {
 
+    private final MechSummaryCache mechSummaryCache;
+
+    public RestoreUnitAction() {
+        this(MechSummaryCache.getInstance());
+    }
+
+    public RestoreUnitAction(MechSummaryCache mechSummaryCache) {
+        this.mechSummaryCache = Objects.requireNonNull(mechSummaryCache);
+    }
+
     @Override
     public void execute(Campaign campaign, Unit unit) {
+        Entity newEntity = getCopyOfEntity(unit.getEntity());
+        if (newEntity != null) {
+            restoreUnit(campaign, unit, newEntity);
+        } else {
+            // Fall back to the old way of restoring a unit if we could not
+            // create a copy of the entity from the summary cache
+            oldUnitRestoration(campaign, unit);
+        }
+
+        MekHQ.triggerEvent(new UnitChangedEvent(unit));
+    }
+
+    /**
+     * Restore a unit by swapping out its entity and replacing its parts.
+     * @param campaign The campaign which owns the unit.
+     * @param unit The unit to restore.
+     * @param newEntity The new entity to assign to the unit.
+     */
+    private void restoreUnit(Campaign campaign, Unit unit, Entity newEntity) {
+        // CAW: this logic is broadly similar to Campaign::addNewUnit
+        final Entity oldEntity = unit.getEntity();
+        newEntity.setId(oldEntity.getId());
+
+        campaign.getGame().removeEntity(oldEntity.getId(), 0);
+
+        newEntity.setOwner(campaign.getPlayer());
+        newEntity.setGame(campaign.getGame());
+        newEntity.setExternalIdAsString(unit.getId().toString());
+        campaign.getGame().addEntity(newEntity.getId(), newEntity);
+
+        unit.setEntity(newEntity);
+
+        unit.removeParts();
+
+        unit.initializeBaySpace();
+
+        unit.initializeParts(true);
+        unit.runDiagnostic(false);
+        unit.setSalvage(false);
+        unit.resetPilotAndEntity();
+    }
+
+    /**
+     * Get a copy of the entity from the {@link MechSummaryCache}.
+     * @param en The entity to copy.
+     * @return A copy of the entity, or {@code null} if a copy could not be made.
+     */
+    @Nullable
+    private Entity getCopyOfEntity(Entity en) {
+        final MechSummary ms = mechSummaryCache.getMech(en.getShortNameRaw());
+        try {
+            if (ms != null) {
+                return new MechFileParser(ms.getSourceFile(), ms.getEntryName()).getEntity();
+            }
+        } catch (EntityLoadingException e) {
+            MekHQ.getLogger().error("Cannot restore unit from entity, could not find: " + ms.getName(), e);
+        }
+
+        return null;
+    }
+
+    /**
+     * Restores a unit using the old per-part logic.
+     * @param campaign The campaign which owns the unit.
+     * @param unit The unit to restore.
+     */
+    private void oldUnitRestoration(Campaign campaign, Unit unit) {
+        MekHQ.getLogger().warning("Falling back to old unit restoration logic");
+
         unit.setSalvage(false);
 
         boolean needsCheck = true;
@@ -105,7 +186,5 @@ public class RestoreUnitAction implements IUnitAction {
                 }
             }
         }
-
-        MekHQ.triggerEvent(new UnitChangedEvent(unit));
     }
 }

--- a/MekHQ/src/mekhq/campaign/unit/actions/RestoreUnitAction.java
+++ b/MekHQ/src/mekhq/campaign/unit/actions/RestoreUnitAction.java
@@ -36,19 +36,28 @@ import mekhq.campaign.unit.Unit;
  */
 public class RestoreUnitAction implements IUnitAction {
 
-    private final MechSummaryCache mechSummaryCache;
+    private final IEntityCopyFactory entityCopyFactory;
 
+    /**
+     * Creates a new {@code RestoreUnitAction} instance using
+     * the default means of creating entity copies.
+     */
     public RestoreUnitAction() {
-        this(MechSummaryCache.getInstance());
+        this(new FileSystemEntityCopyFactory());
     }
 
-    public RestoreUnitAction(MechSummaryCache mechSummaryCache) {
-        this.mechSummaryCache = Objects.requireNonNull(mechSummaryCache);
+    /**
+     * Creates a new {@code RestoreUnitAction} instance using
+     * the provided {@link IEntityCopyFactory}.
+     * @param entityCopyFactory The factory to create entity copies with.
+     */
+    public RestoreUnitAction(IEntityCopyFactory entityCopyFactory) {
+        this.entityCopyFactory = Objects.requireNonNull(entityCopyFactory);
     }
 
     @Override
     public void execute(Campaign campaign, Unit unit) {
-        Entity newEntity = getCopyOfEntity(unit.getEntity());
+        Entity newEntity = entityCopyFactory.copy(unit.getEntity());
         if (newEntity != null) {
             restoreUnit(campaign, unit, newEntity);
         } else {
@@ -88,25 +97,6 @@ public class RestoreUnitAction implements IUnitAction {
         unit.runDiagnostic(false);
         unit.setSalvage(false);
         unit.resetPilotAndEntity();
-    }
-
-    /**
-     * Get a copy of the entity from the {@link MechSummaryCache}.
-     * @param en The entity to copy.
-     * @return A copy of the entity, or {@code null} if a copy could not be made.
-     */
-    @Nullable
-    private Entity getCopyOfEntity(Entity en) {
-        final MechSummary ms = mechSummaryCache.getMech(en.getShortNameRaw());
-        try {
-            if (ms != null) {
-                return new MechFileParser(ms.getSourceFile(), ms.getEntryName()).getEntity();
-            }
-        } catch (EntityLoadingException e) {
-            MekHQ.getLogger().error("Cannot restore unit from entity, could not find: " + ms.getName(), e);
-        }
-
-        return null;
     }
 
     /**
@@ -185,6 +175,44 @@ public class RestoreUnitAction implements IUnitAction {
                     }
                 }
             }
+        }
+    }
+
+    /**
+     * Implementations allow copies of entities to be created.
+     */
+    public interface IEntityCopyFactory {
+        /**
+         * Gets a copy of the entity.
+         * @param en The entity to copy.
+         * @return A copy of the entity, or {@code null} if a copy could not be made.
+         */
+        @Nullable
+        public Entity copy(Entity entity);
+    }
+
+    /**
+     * Gets a copy of the entity from the file system, via {@link MechSummaryCache}
+     * and {@link MechFileParser}.
+     */
+    private static class FileSystemEntityCopyFactory implements IEntityCopyFactory {
+        /**
+         * Get a copy of the entity from the {@link MechSummaryCache}.
+         * @param en The entity to copy.
+         * @return A copy of the entity, or {@code null} if a copy could not be made.
+         */
+        @Nullable
+        public Entity copy(Entity en) {
+            final MechSummary ms = MechSummaryCache.getInstance().getMech(en.getShortNameRaw());
+            try {
+                if (ms != null) {
+                    return new MechFileParser(ms.getSourceFile(), ms.getEntryName()).getEntity();
+                }
+            } catch (EntityLoadingException e) {
+                MekHQ.getLogger().error("Cannot restore unit from entity, could not find: " + ms.getName(), e);
+            }
+
+            return null;
         }
     }
 }

--- a/MekHQ/unittests/mekhq/campaign/unit/actions/RestoreUnitActionTest.java
+++ b/MekHQ/unittests/mekhq/campaign/unit/actions/RestoreUnitActionTest.java
@@ -18,8 +18,9 @@
  */
 package mekhq.campaign.unit.actions;
 
-import static org.junit.Assert.*;
 import static org.mockito.Mockito.*;
+
+import java.util.UUID;
 
 import mekhq.campaign.Campaign;
 import mekhq.campaign.Quartermaster;
@@ -32,14 +33,69 @@ import mekhq.campaign.unit.Unit;
 import org.junit.Test;
 
 import megamek.common.Entity;
+import megamek.common.Game;
+import megamek.common.Player;
 
 public class RestoreUnitActionTest {
     @Test
-    public void restoreUnitDoesNotSkipPartsAfterRepairingMissingParts() {
+    public void restoreUnitSwitchesOutEntityAndParts() {
+        Campaign mockCampaign = mock(Campaign.class);
+        Game mockGame = mock(Game.class);
+        when(mockCampaign.getGame()).thenReturn(mockGame);
+        Player mockPlayer = new Player(1, "Player");
+        when(mockCampaign.getPlayer()).thenReturn(mockPlayer);
+        Quartermaster mockQuartermaster = mock(Quartermaster.class);
+        when(mockCampaign.getQuartermaster()).thenReturn(mockQuartermaster);
+        
+        int entityId = 42;
+        String shortName = "Test Mech TST-01X";
+        Entity mockEntity = mock(Entity.class);
+        when(mockEntity.getId()).thenReturn(entityId);
+        when(mockEntity.getShortNameRaw()).thenReturn(shortName);
+        
+        UUID id = UUID.randomUUID();
+        Unit unit = mock(Unit.class);
+        when(unit.getId()).thenReturn(id);
+        when(unit.getEntity()).thenReturn(mockEntity);
+
+        Entity mockNewEntity = mock(Entity.class);
+        doAnswer(inv -> {
+            int newId = (int) inv.getArgument(0);
+            when(mockNewEntity.getId()).thenReturn(newId);
+            return null;
+        }).when(mockNewEntity).setId(anyInt());
+        RestoreUnitAction.IEntityCopyFactory mockEntityCopyFactory = mock(RestoreUnitAction.IEntityCopyFactory.class);
+        doReturn(mockNewEntity).when(mockEntityCopyFactory).copy(eq(mockEntity));
+
+        RestoreUnitAction action = new RestoreUnitAction(mockEntityCopyFactory);
+
+        action.execute(mockCampaign, unit);
+
+        verify(mockNewEntity, times(1)).setId(eq(entityId));
+        verify(mockNewEntity, times(1)).setOwner(eq(mockPlayer));
+        verify(mockNewEntity, times(1)).setGame(eq(mockGame));
+        verify(mockNewEntity, times(1)).setExternalIdAsString(eq(id.toString()));
+        
+        verify(mockGame, times(1)).removeEntity(eq(entityId), anyInt());
+        verify(mockGame, times(1)).addEntity(eq(entityId), eq(mockNewEntity));
+
+        verify(unit, times(1)).setEntity(eq(mockNewEntity));
+        verify(unit, times(1)).removeParts();
+        verify(unit, times(1)).initializeBaySpace();
+        verify(unit, times(1)).initializeParts(eq(true));
+        verify(unit, times(1)).runDiagnostic(eq(false));
+        verify(unit, times(1)).setSalvage(eq(false));
+        verify(unit, times(1)).resetPilotAndEntity();
+    }
+
+    @Test
+    public void restoreUnitUsingOldStrategy() {
+        RestoreUnitAction.IEntityCopyFactory mockEntityCopyFactory = mock(RestoreUnitAction.IEntityCopyFactory.class);
         Campaign mockCampaign = mock(Campaign.class);
         Quartermaster mockQuartermaster = mock(Quartermaster.class);
         when(mockCampaign.getQuartermaster()).thenReturn(mockQuartermaster);
         Entity mockEntity = mock(Entity.class);
+        when(mockEntity.getShortNameRaw()).thenReturn("Test Mech TST-01X");
         Unit unit = new Unit(mockEntity, mockCampaign);
 
         Part part0 = mock(Part.class);
@@ -70,7 +126,7 @@ public class RestoreUnitActionTest {
         unit.addPart(missing1);
         unit.addPart(part1);
 
-        RestoreUnitAction action = new RestoreUnitAction();
+        RestoreUnitAction action = new RestoreUnitAction(mockEntityCopyFactory);
 
         action.execute(mockCampaign, unit);
 

